tests: Add initial test case for Modbus Server Container

mgeeIOL opened the pull request:
# Description

Includes new test group for the modbus server, alongside the initial test case where the modbus server is coneceted to by a client on the github action's runner. Currently, the github actions runner connects to the modbus server to read and write to the LED and Button register(s). Will need to update test case as registers are implemented in the modbus server container.

This commit also includes a refactoring of parts of build.yml in the github actions to allow the actions runner to flash the board with the modbus server. Also changes the overall workflow slighly in order to make future container testing on the b_u585i_iot02a board easier to implement.

## Type of change

Please delete options that are not relevant.

- [x] CI system update
- [x] Test Coverage update

# How Has This Been Tested?

Github actions changes will be tested when the PR is opened, given that it is a change to how the GHA workflows are set up.

Modbus server tests have been ran manually over a serial connection to the board with the appropriate container flashed. 

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

PatrickRobbIOL commented on .github/workflows/build.yml:
```diff
@@ -1,4 +1,7 @@
 name: Build
+concurrency:
+  group: pr-workflows
+  cancel-in-progress: false
```
It's worth explaining this change in the commit message (we need to run workflows for different PRs in serial instead of synchronous since they are sharing a common resource).

PatrickRobbIOL commented on .github/workflows/build.yml:
```diff
@@ -127,37 +130,31 @@ jobs:
         sample:
-          - name: hello-world
-            expected: "powered by Ocre"
+          - name: generic-hello-world
```
I thought changing the name of the hello world artifact was out of scope but I guess it's okay given that we are extending usage of board specific container image (thus it makes sense to clarify what is board specific vs generic). So, I guess no changes needed.

PatrickRobbIOL reviewed (dismissed):
I left a few minor comments but it looks good. Approving this PR with the understanding that Matt will push an update per my comments. What do you think @kr-t ?

SorinOlari commented on .github/workflows/build.yml:
```diff
@@ -172,9 +166,31 @@ jobs:
+      - name: Setup Zephyr project
+        uses: zephyrproject-rtos/action-zephyr-setup@v1
+        with:
+          app-path: application
```
`+`

SorinOlari commented on .github/workflows/build.yml:
```diff
@@ -189,13 +205,69 @@ jobs:
+        with:
+          path: application
+
+      - name: Clone ocre-sdk
```
We are downloading .wasm file from artifacts, so ocre-sdk is unused here

SorinOlari commented on .github/workflows/build.yml:
```diff
@@ -223,22 +295,38 @@ jobs:
-          # Add here more samples
-    steps:
+    steps: 
+      - name: Install tools (xxd + WASI SDK)
```
the same thing as in `build-and-run-linux-sample`

SorinOlari commented on .github/workflows/build.yml:
```diff
@@ -344,3 +429,107 @@ jobs:
+        run: |
+          git submodule update --init --recursive
+
+      - name: Install tools (xxd + WASI SDK)
```
the same thing as above

SorinOlari commented on .github/workflows/build.yml:
```diff
@@ -344,3 +429,107 @@ jobs:
 
+
+  build-zephyr-modbus_server-b_u585i_iot02a:
+    needs: [build-wasm-files]
```
remove square brackets please

SorinOlari commented:
I added some comments, I hope my theory is correct)
I think it would also be better not to save the .wasm files until the end, could you try deleting them after we used them?
Another option would be if we could save them all in a single artifact.
You could just try, until I think it's not a priority if that isn't part of the task @kr-t ?

mgeeIOL commented:
I have updated the PR to include the optimization changes commented on by @SorinOlari , let me know if I missed anything. 

The logging artifacts are removed per the comment from @kr-t. As for removing the wasm artifacts I modified the workflow to have every wasm file be a part of the same artifact by saving them locally and uploading the directory as one artifact in a separate step. 

This is the most practical option we have for handling the wasm artifacts as there is no official way to delete github artifacts and some jobs of the workflow require the wasm files and are not run on the zephyr-xlarge-runner so we cannot just store them locally. 
 
To my knowledge the current artifacts are as close as we can get to the intention described earlier with the current workflow setup, if anything needs to change let me know.
